sonar-scm-svn-plugin/src/main/java/org/sonar/plugins/scm/svn/ChangedLinesComputer.java

```diff
@@ -0,0 +1,163 @@
+  };
+
+  ChangedLinesComputer(Set<Path> included) {
+    this.tracker = new Tracker(included);
+  }
+
+  OutputStream receiver() {
+    return receiver;
+  }
+
+  /**
+   * From a stream of svn-style unified diff lines,
+   * compute the line numbers that should be considered changed.
+   *
+   * Example input:
+   * <pre>
+   * Index: path/to/file
+   * ===================================================================
+   * --- lao 2002-02-21 23:30:39.942229878 -0800
+   * +++ tzu 2002-02-21 23:30:50.442260588 -0800
+   * @@ -1,7 +1,6 @@
+   * -The Way that can be told of is not the eternal Way;
+   * -The name that can be named is not the eternal name.
+   *  The Nameless is the origin of Heaven and Earth;
+   * -The Named is the mother of all things.
+   * +The named is the mother of all things.
+   * +
+   *  Therefore let there always be non-being,
+   *    so we may see their subtlety,
+   *  And let there always be being,
+   * @@ -9,3 +8,6 @@
+   *  The two are the same,
+   *  But after they are produced,
+   *    they have different names.
+   * +They both may be called deep and profound.
+   * +Deeper and more profound,
+   * +The door of all subtleties!
+   * </pre>
+   *
+   * See also: http://www.gnu.org/software/diffutils/manual/html_node/Example-Unified.html#Example-Unified
+   */
+  Map<Path, Set<Integer>> changedLines() {
+    return tracker.changedLines();
+  }
+
+  private static class Tracker {
+
+    private static final Pattern START_LINE_IN_TARGET = Pattern.compile(" \\+(\\d+)");
+    private static final String ENTRY_START_PREFIX = "Index: ";
+
+    private final Map<Path, Set<Integer>> changedLines = new HashMap<>();
+    private final Set<Path> included;
+
+    private int lineNumInTarget;
+    private Path currentPath = null;
+    private int skipCount = 0;
+
+    Tracker(Set<Path> included) {
+      this.included = included;
+    }
+
+    private void parseContent(String line) {
+      char firstChar = line.charAt(0);
+      if (firstChar == ' ') {
+        lineNumInTarget++;
+      } else if (firstChar == '+') {
+        changedLines.get(currentPath).add(lineNumInTarget);
+        lineNumInTarget++;
+      }
+    }
+
+    private void parseLine(String line) {
+      if (line.startsWith(ENTRY_START_PREFIX)) {
+        removeCurrentChangesetIfEmpty();
+        currentPath = Paths.get(line.substring(ENTRY_START_PREFIX.length()).trim());
+        if (!included.contains(currentPath)) {
+          return;
+        }
+        changedLines.put(currentPath, new HashSet<>());
+        skipCount = 3;
+        return;
+      }
+
+      if (!included.contains(currentPath)) {
+        return;
+      }
+
+      if (skipCount > 0) {
+        skipCount--;
+        return;
+      }
+
+      if (line.startsWith("@@ ")) {
+        Matcher matcher = START_LINE_IN_TARGET.matcher(line);
+        if (!matcher.find()) {
+          throw new IllegalStateException();
+        }
+        lineNumInTarget = Integer.parseInt(matcher.group(1));
```

Comment:
This can throw an ugly unchecked exception

Reply:
I think this will always succeed, because the matched group is defined as `(\d+)`